jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java

```diff
@@ -40,23 +50,24 @@
  */
 public class StepsRunner {
 
-  /** Holds the individual steps. */
-  private static class Steps {
+  /** Holds the individual step results. */
+  private static class StepResults {
 
-    @Nullable private RetrieveRegistryCredentialsStep retrieveTargetRegistryCredentialsStep;
-    @Nullable private AuthenticatePushStep authenticatePushStep;
-    @Nullable private PullBaseImageStep pullBaseImageStep;
-    @Nullable private PullAndCacheBaseImageLayersStep pullAndCacheBaseImageLayersStep;
-
-    @Nullable
-    private ImmutableList<BuildAndCacheApplicationLayerStep> buildAndCacheApplicationLayerSteps;
-
-    @Nullable private PushLayersStep pushBaseImageLayersStep;
-    @Nullable private PushLayersStep pushApplicationLayersStep;
-    @Nullable private BuildImageStep buildImageStep;
-    @Nullable private PushContainerConfigurationStep pushContainerConfigurationStep;
+    private static <E> Future<E> failedFuture() {
+      return Futures.immediateFailedFuture(
+          new IllegalStateException("invalid usage; required step not configured"));
+    }
 
-    @Nullable private AsyncStep<BuildResult> finalStep;
+    private Future<ImageAndAuthorization> baseImageAndAuth = failedFuture();
+    private Future<List<Future<CachedLayerAndName>>> baseImageLayers = failedFuture();
+    private List<Future<CachedLayerAndName>> applicationLayers = new ArrayList<>();
+    private Future<Image> builtImage = failedFuture();
+    private Future<Credential> targetRegistryCredentials = failedFuture();
+    private Future<Authorization> pushAuthorization = failedFuture();
+    private Future<List<Future<BlobDescriptor>>> baseImageLayerPushResults = failedFuture();
+    private Future<List<Future<BlobDescriptor>>> applicationLayerPushResults = failedFuture();
```

Comment:
Note to reviewers: layers push results (digest and size of pushed layers) are not actually used in the both old and new designs. They are merely for blocking the final registry push step until we are certain that the layers are all pushed.